doc/design/s3-undelete.md

```diff
@@ -0,0 +1,69 @@
+
+Our ultimate goal is to prevent data loss from application programming errors. With protection such as an undelete capability, we will be safer in implementing application features that involve intentional deletion of data. Any bugs we introduce while doing so are far less likely to destroy data that was not supposed to be deleted.
+
+The original GitHub issue around this feature request can be found at [https://github.com/dandi/dandi-archive/issues/524](https://github.com/dandi/dandi-archive/issues/524). Although the issue asks for a Deep Glacier storage tier, the design in this document solves the underlying problem differently (and in a more robust way). Below we address the possible usage of a Deep Glacier tiered bucket as a solution to the orthogonal problem of data ******backup******, which addresses a different problem than the undelete capability described in this document.
+
+## **Requirements**
+
+- After deletion of an asset blob, there needs to be a period of 30 days during which that blob can be restored.
+
+## **Proposed Solution**
+
+What we want can be described as a “trailing delete” mechanism. Upon deletion of an asset from the bucket, we would like the object to remain recoverable for some amount of time. S3 already supports this in the form of Bucket Versioning.
+
+### **S3 Bucket Versioning**
+
+Enabling bucket versioning will change what happens when an object in S3 is deleted. Instead of permanently deleting the object, S3 will simply place a delete marker on it. At that point, the object is hidden from view and appears to be deleted, but still exists and is recoverable.
+
+In addition, we can place an S3 Lifecycle policy on the bucket that automatically clears delete markers and “permanently deletes” their associated objects after some set amount of time.
+
+```
+  # Terraform-encoded lifecycle rule.
+  # Based on https://docs.aws.amazon.com/AmazonS3/latest/userguide/lifecycle-configuration-examples.html#lifecycle-config-conceptual-ex7
+  rule {
+    id = "ExpireOldDeleteMarkers"
+    filter {}
+
+    # Expire objects with delete markers after 1 day
+    noncurrent_version_expiration {
+      noncurrent_days = 1
+    }
+
+    # Also delete any delete markers associated with the expired object
+    expiration {
+      expired_object_delete_marker = true
+    }
```

Comment:
oh, that is neat! I didn't know that is possible.
Just to make sure: Would it be in effect also for old/already existing DeleteMarkers or somehow works only on the new ones which it detects being placed after adding this lifecycle rule?

Reply:
It would apply to existing delete markers as well. The docs mention this briefly, and I also tested it myself to confirm.

https://aws.amazon.com/s3/faqs/
> S3 Lifecycle policies apply to both existing and new S3 objects, helping you optimize storage and maximize cost savings for all current data and any new data placed in S3 without time-consuming manual data review and migration.